contrib/etcd/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,158 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerAgentName = "resource-controller-adapter"
+	updateEventType     = "dev.knative.etcd.object.update"
+	deleteEventType     = "dev.knative.etcd.object.delete"
+
+	envSinkURI    = "SINK_URI"
+	envAPIVersion = "APIVERSION_"
+	envKind       = "KIND_"
+)
+
+// Add creates new EtcdSource controllers, one per resource,
+// and adds them to the manager.
+func Add(mgr manager.Manager) error {
+	sinkURI, defined := os.LookupEnv(envSinkURI)
+	if !defined {
+		return fmt.Errorf("required environment variable not defined '%s'", envSinkURI)
+	}
+
+	eventsClient, err := kncloudevents.NewDefaultClient(sinkURI)
+	if err != nil {
+		return err
+	}
+
+	// Create one controller per watch
+	i := 0
+	for {
+		apiVersion, defined := os.LookupEnv(envAPIVersion + strconv.Itoa(i))
+		if !defined {
+			break
+		}
+
+		kind, defined := os.LookupEnv(envKind + strconv.Itoa(i))
+		if !defined {
+			return fmt.Errorf("required environment variable not defined '%s'", envKind+strconv.Itoa(i))
+		}
+		gv, err := schema.ParseGroupVersion(apiVersion)
+		if err != nil {
+			return err
+		}
+
+		// don't use the sdk here as we are using unstructured watches
+		err = add(mgr, &reconciler{
+			client:       mgr.GetClient(),
+			eventsClient: eventsClient,
+			gvk:          schema.GroupVersionKind{Kind: kind, Group: gv.Group, Version: gv.Version},
+		})
+
+		if err != nil {
+			return err
+		}
+
+		i++
+	}
+	return nil
+}
+
+// Add returns a controller for esource-controller-adapter.
+func add(mgr manager.Manager, r *reconciler) error {
+	// Setup a new controller to Reconcile Subscriptions.
+	c, err := controller.New(controllerAgentName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch Parent events and enqueue Parent object key.
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(r.gvk)
+
+	err = c.Watch(&source.Kind{Type: u}, &handler.EnqueueRequestForObject{})
+	return nil
+}
+
+// reconciler reconciles a EtcdSource object
+type reconciler struct {
+	client       client.Client
+	eventsClient eventsclient.Client
+	gvk          schema.GroupVersionKind
+}
+
+func (r *reconciler) InjectClient(c client.Client) error {
+	r.client = c
+	return nil
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.TODO()
+	logger := logging.FromContext(ctx)
+
+	object := &unstructured.Unstructured{}
+	object.SetGroupVersionKind(r.gvk)
+	err := r.client.Get(ctx, request.NamespacedName, object)
+
+	if errors.IsNotFound(err) {
+		logger.Errorf("could not find %s %v\n", request.Name, request)
+		return reconcile.Result{}, nil
+	}
+	eventType := updateEventType
+	if object.GetDeletionTimestamp() != nil {
+		eventType = deleteEventType
+	}
+
+	event := cloudevents.Event{
+		Context: cloudevents.EventContextV03{
```

Comment:
V03 is not released yet (the cloudevents spec is still changing). Please use V02 for now.